src/libsystemd/sd-device/device-enumerator.c

```diff
@@ -278,16 +299,99 @@ static int device_compare(sd_device * const *_a, sd_device * const *_b) {
+        sd_device **devices;
+        sd_device *device;
+        size_t n = 0, m;
+        int r;
+
+        assert(enumerator);
+
+        devices = new(sd_device*, hashmap_size(enumerator->devices_by_syspath));
+        if (!devices)
+                return -ENOMEM;
+
+        if (!set_isempty(enumerator->prioritized_subsystem)) {
+                const char *syspath;
+
+                HASHMAP_FOREACH_KEY(device, syspath, enumerator->devices_by_syspath) {
+                        _cleanup_free_ char *p = NULL;
+                        const char *subsys;
+
+                        if (sd_device_get_subsystem(device, &subsys) < 0)
+                                continue;
+
+                        if (!set_contains(enumerator->prioritized_subsystem, subsys))
+                                continue;
+
+                        devices[n++] = sd_device_ref(device);
+
+                        for (;;) {
+                                _cleanup_free_ char *q = NULL;
+
+                                r = path_extract_directory(p ?: syspath, &q);
+                                if (r == -EADDRNOTAVAIL)
+                                        break;
+                                if (r < 0)
+                                        goto failed;
+
+                                device = hashmap_get(enumerator->devices_by_syspath, q);
+                                if (device)
+                                        devices[n++] = sd_device_ref(device);
+
+                                free_and_replace(p, q);
+                        }
+                }
+
+                for (size_t i = 0; i < n; i++) {
+                        r = sd_device_get_syspath(devices[i], &syspath);
+                        if (r < 0)
+                                goto failed;
+
+                        sd_device_unref(hashmap_remove(enumerator->devices_by_syspath, syspath));
+                }
+
+                typesafe_qsort(devices, n, device_compare);
+        }
+
+        m = n;
+        while ((device = hashmap_steal_first(enumerator->devices_by_syspath)))
+                devices[n++] = device;
+
+        typesafe_qsort(devices + m, n - m, device_compare);
+
+        enumerator->n_devices = n;
+        free_and_replace(enumerator->devices, devices);
+        return 0;
+
+failed:
+        for (size_t i = 0; i < n; i++)
+                sd_device_unref(devices[i]);
+
+        free(devices);
+        return r;
+}
+
 int device_enumerator_add_device(sd_device_enumerator *enumerator, sd_device *device) {
+        const char *syspath;
+        int r;
+
         assert_return(enumerator, -EINVAL);
         assert_return(device, -EINVAL);
 
-        if (!GREEDY_REALLOC(enumerator->devices, enumerator->n_devices + 1))
-                return -ENOMEM;
+        r = sd_device_get_syspath(device, &syspath);
+        if (r < 0)
+                return r;
+
+        r = hashmap_ensure_put(&enumerator->devices_by_syspath, &string_hash_ops, syspath, device);
```

Comment:
string_hash_ops → path_hash_ops?

Reply:
hmm, shouldn't this be an ordered hashmap?

Reply:
I initially thought here path_hash_ops should be used, but syspath is always normalized, so it is not necessary, and string_hash_ops is faster.

I cannot find no reason that ordered hashmap should be used here... Why do you think so?

Reply:
oh, mostly for illustrative purposes, i.e. it makes clear what the contents here is. I mean, does the performance really matter here?

Also, I'd consider it preparation for the future. Given we now have this indexed already, maybe sooner or later we'll expose this, i.e. add sd_device_enumerator_get_device_by_syspath() or so, and once we do that we cannot rely on normalization of the passed path.

but dunno... doesn't really matter, we could of course still change this later if needed.